File: day25.rs
use std::io;

fn input() -> Vec<Vec<char>> {
    let mut grid = Vec::new();

    let mut done = false;
    while !done {
        let mut buf = String::new();
        match io::stdin().read_line(&mut buf) {
            Ok(..) => {
                if buf.len() == 0 {
                    done = true;
                } else {
                    grid.push(buf.trim().chars().collect());
                }
            }
            Err(..) => {
                done = true;
            }
        }
    }

    return grid;
}

fn step_right(grid: &mut Vec<Vec<char>>) -> bool {
    let n = grid.len();
    let m = grid[0].len();

    let mut cpy = Vec::new();
    for i in 0..n {
        cpy.push(grid[i].clone());
    }

    let mut upd = false;
    for i in 0..n {
        for j in 0..m {
            if grid[i][j] == '>' && cpy[i][(j + 1) % m] == '.' {
                upd = true;
                grid[i][j] = '.';
                grid[i][(j + 1) % m] = '*';
            }
        }
    }
    for i in 0..n {
        for j in 0..m {
            if grid[i][j] == '*' {
                grid[i][j] = '>';
            }
        }
    }
    return upd;
}

fn step_down(grid: &mut Vec<Vec<char>>) -> bool {
    let n = grid.len();
    let m = grid[0].len();

    let mut cpy = Vec::new();
    for i in 0..n {
        cpy.push(grid[i].clone());
    }

    let mut upd = false;
    for i in 0..n {
        for j in 0..m {
            if grid[i][j] == 'v' && cpy[(i + 1) % n][j] == '.' {
                upd = true;
                grid[i][j] = '.';
                grid[(i + 1) % n][j] = '*';
            }
        }
    }
    for i in 0..n {
        for j in 0..m {
            if grid[i][j] == '*' {
                grid[i][j] = 'v';
            }
        }
    }
    return upd;
}

fn step(grid: &mut Vec<Vec<char>>) -> bool {
    let r = step_right(grid);
    let d = step_down(grid);
    return r || d;
}

fn main() {
    let mut grid = input();
    let mut i = 0;
    while step(&mut grid) {
        i += 1;
    }
    println!("{}", i + 1);
}

File: day25.go
package main

import (
	"bufio"
	"bytes"
	"encoding/gob"
	"fmt"
	"os"
)

func deepcopy(out, in interface{}) {
	buf := new(bytes.Buffer)
	gob.NewEncoder(buf).Encode(in)
	gob.NewDecoder(buf).Decode(out)
}

func stepAlong(grid [][]byte, c byte, di, dj int) bool {
	n := len(grid)
	m := len(grid[0])
  var cpy [][]byte
	deepcopy(&cpy, grid)
	upd := false

	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if grid[i][j] == c && cpy[(i+di)%n][(j+dj)%m] == '.' {
				upd = true
				grid[i][j] = '.'
				grid[(i+di)%n][(j+dj)%m] = '*'
			}
		}
	}
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if grid[i][j] == '*' {
				grid[i][j] = c
			}
		}
	}
	return upd
}

func step(grid [][]byte) bool {
	d := stepAlong(grid, '>', 0, 1)
	r := stepAlong(grid, 'v', 1, 0)
	return d || r
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	var grid [][]byte
	for scanner.Scan() {
		line := scanner.Text()
		row := make([]byte, len(line))
		for j := range row {
			row[j] = line[j]
		}
		grid = append(grid, row)
	}

	ans := 1
	for step(grid) {
		ans++
	}
	fmt.Println(ans)
}
